src/System.Device.Gpio/System/Device/I2c/UnixI2cBus.cs

```diff
@@ -0,0 +1,218 @@
+    {
+        private const string DefaultDevicePath = "/dev/i2c";
+        private static readonly object s_initializationLock = new object();
+        public int BusId { get; }
+        protected int BusFileDescriptor { get; private set; }
+        private HashSet<int>? _usedAddresses = null;
+
+        public static new unsafe UnixI2cBus Create(int busId)
+        {
+            string deviceFileName = $"{DefaultDevicePath}-{busId}";
+            lock (s_initializationLock)
+            {
+                int busFileDescriptor = Interop.open(deviceFileName, FileOpenFlags.O_RDWR);
+
+                if (busFileDescriptor < 0)
+                {
+                    throw new IOException($"Error {Marshal.GetLastWin32Error()}. Can not open I2C device file '{deviceFileName}'.");
+                }
+
+                I2cFunctionalityFlags functionalityFlags;
+                int result = Interop.ioctl(busFileDescriptor, (uint)I2cSettings.I2C_FUNCS, new IntPtr(&functionalityFlags));
+                if (result < 0)
+                {
+                    functionalityFlags = 0;
+                }
+
+                if ((functionalityFlags & I2cFunctionalityFlags.I2C_FUNC_I2C) != 0)
+                {
+                    return new UnixI2cBus(busFileDescriptor, busId);
+                }
+                else
+                {
+                    return new UnixI2cFileTransferBus(busFileDescriptor, busId);
+                }
+            }
+        }
+
+        public UnixI2cBus(int busFileDescriptor, int busId)
+        {
+            BusId = busId;
+            BusFileDescriptor = busFileDescriptor;
+        }
+
+        public override I2cDevice CreateDevice(int deviceAddress)
+        {
+            if (!HasValidFileDescriptor())
+            {
+                throw new ObjectDisposedException(nameof(UnixI2cBus));
+            }
+
+            _usedAddresses ??= new HashSet<int>();
+            if (!_usedAddresses.Add(deviceAddress))
+            {
+                throw new ArgumentException($"Device with address 0x{deviceAddress,0X2} is already open.", nameof(deviceAddress));
+            }
+
+            return CreateDeviceNoCheck(deviceAddress);
+        }
+
+        internal I2cDevice CreateDeviceNoCheck(int deviceAddress)
+        {
+            return new UnixI2cDevice(this, deviceAddress);
+        }
+
+        public override void RemoveDevice(int deviceAddress)
+        {
+            if (!RemoveDeviceNoCheck(deviceAddress))
+            {
+                throw new ArgumentException($"Device with address 0x{deviceAddress,0X2} was not open.", nameof(deviceAddress));
+            }
+        }
+
+        internal bool RemoveDeviceNoCheck(int deviceAddress)
+        {
+            return _usedAddresses?.Remove(deviceAddress) ?? false;
+        }
+
+        public override unsafe void Read(int deviceAddress, Span<byte> buffer)
+        {
+            if (deviceAddress < 0 || deviceAddress > ushort.MaxValue)
```

Comment:
for I2C the max address is 2014 or 128 but support for 128+ address space is not supported (yet) in .NET IoT. See https://www.i2c-bus.org/addressing/. And for 10 bits addresses here: https://www.i2c-bus.org/addressing/10-bit-addressing/ and that's something that can be done on a higher level.
So max value checked should be either 128 or 1024 or just 256 to be safe and allow the 1024 scenario to happen on a higher level.

Reply:
I did ushort.MaxValue because that's the max we can pass to the native code. I prefer on letting native code handle the range part as there may be some more scenarios we're missing here and native can handle.

Reply:
OK, makes sense then.